### ISSUE TITLE
Abbreviated single argument function literals

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#265]
This is a proposal to add support for _abbreviated single-argument function literals_ to Dart.

It is based on discussions about [anonymous methods](#260), and may be used in combination with a pipe operator to achieve expressions of a similar form as an anonymous method. A similar mechanism exists [in Kotlin](https://kotlinlang.org/docs/reference/lambdas.html#it-implicit-name-of-a-single-parameter).

## Overview

Function literals are used frequently in Dart, and many of them declare exactly one required parameter and no optional parameters. The declaration of such a parameter can be omitted if a standard name is chosen for it, and if a suitable parameter type can be obtained from type inference.

Syntactically, a block function literal with no parameters already has the form `() { ... }`, so we cannot just omit the parameter declaration in order to specify that the parameter should be declared implicitly. A plain `<block>` could be used (that is `{ ... }` containing statements), but this could create additional ambiguities with the expression `{}` (which might already be a set literal or a map literal). So we require that it is a block that contains one or more statements: `<nonEmptyBlock>`.

It is already a compile-time error to have an expression statement starting with `{`, so there is no ambiguity for a block in a sequence of statements: That is just some more statements in a nested static scope, it is not a function literal that we will evaluate and ignore.

For function literals of the form `(it) => e`, we could use the abbreviation `=> e`. In this case there are no syntactic conflicts.

Following Kotlin and using `it` as the default parameter name, we get the following:

```dart
main() {
  var xs = [1, 2, 3];
  // Current form.
  xs.forEach((x) { print(x); });
  // Proposed abbreviation.
  xs.forEach({ print(it); });
  xs.forEach(=> print(it));
}
```

Asynchronous and generator variants can be expressed by adding the relevant keyword in front, e.g., `[2, 3].map(sync* { yield it; yield it; })`.

## Syntax

The grammar is adjusted as follows in order to support abbreviated single-argument function literals:

```
<functionExpression> ::= // Modified (adding ?)
    <formalParameterPart>? <functionExpressionBody>

<functionExpressionWithoutCascade> ::= // Modified (adding ?)
    <formalParameterPart>? <functionExpressionWithoutCascadeBody>

<functionPrimary> ::= // Modified (adding alternative)
    <formalParameterPart> <functionPrimaryBody>
  | <nonEmptyBlock>

<nonEmptyBlock> ::= // New.
    '{' <statement> <statements> '}'
```

## Static Analysis and Dynamic Semantics

An expression `e` of the form `<nonEmptyBlock>` is treated as `(it) e`. *This determines the static analysis as well as the dynamic semantics of such expressions. Note that we need to insist that it is non-empty, because an empty block can be a set literal or a map literal as well.*

An expression `e` of the form `<functionExpressionBody>` or `<functionExpressionWithoutCascadeBody>` is treated as `(it) e`. *This again determines the static analysis as well as the dynamic semantics of `e`.*

## Discussion

It may be considered confusing to have modifiers like `async` as specified. This is completely optional, and we could just omit the support for doing it. Developers would then have to write an explicit parameter part in the case where they want to use an asynchronous function or a generator function.

**Comment by tatumizer** [comment]
Wanted to react with two thumbs up, but this UI didn't let me :)
FWIW, I find this version much more readable than cascade version (which looks fuzzy with nested initializations).

Just to double check: will the following expression work?
```
var x = { return 42; }
```
Or it should be written differently to work?

One unfortunate thing about dart blocks right now is that when you write
```
foo() {
   { // just a block
     return 0;
   }
}
```
then "return" is interpreted as a return from the foo()! Won't it clash with a new definition?

**Comment by eernstg** [comment]
@tatumizer wrote:
> One unfortunate thing about dart blocks right now

Right, it is a potential source of confusion that `return` in a plain block returns from the enclosing function, and `return` in a function literal is a return from that nested function. But I would actually expect those plain blocks to be rather rare.

**Comment by tatumizer** [comment]
> But I would actually expect those plain blocks to be rather rare.

Absolutely! The only known user currently taking advantage of blocks in dart programs is none other than @lrhn - you can negotiate with him offline. LOL

**Comment by eernstg** [comment]
Negotiating with @lrhn is an ongoing effort. ;-)

**Comment by eernstg** [comment]
After a lot of discussions about the use of the name `this` for the implicit parameter, I've concluded that it it likely to reduce the comprehensibility and readability of the code, so I've removed the parts of the proposal where this idea was discussed.

The most obvious choice for the name of the implicit parameter is then `it`, so that's what the proposal uses now.

On the other hand, @lrhn suggested that we could use the context type to offer two variants: When the context type is a function that takes no arguments a `<nonEmptyBlock>` `b` could desugar to `() b`, and otherwise it desugars to `(it) b`. I've included that idea as well, it seems to be pretty well aligned with other mechanisms in Dart.

**Comment by lrhn** [comment]
Here's a new idea: Any undeclared identifier of the regexp-form `\$+\d+` is implicitly a parameter of a surrounding no-params-specified function. I still want you to write `=>` to specify that it's a function (and where it's delimited) because Dart syntax is very hard to parse without guides.
If you nest functions without parameters, you need to write `$$` and `$$$` as prefix, depending on the nesting.

Then `list.sort(=>$1-$2)` would be a two-argument function because it contains `$1` and `$2` which are not otherwise declared, and `list.fold(=>$2)` is `($1, $2) => $2`. (We can also allow inference to add more parameters than you specify, if it's needed by the context, so `list.fold(=>$1)` is `($1, _) => $1`.

For nested functions, you need to add extra `$`s: `=>=>$1+$$1` is ($1) => ($$1) => $1 + $$1`. 

We can allow you to omit the digit if there is only one parameter, so the above can be `=>=>$ + $$`.
We can allow you to write `$foo` to designate a named parameter named `foo`, so `=>$foo` is `({foo}) => foo`. (We remove the `$`s from the name when you call the function, but internally we ensure disambiguation, so `=>=>$foo+$$foo` is `({foo}) => let $foo = foo in ({foo}) => let $$foo = foo in $foo + $$foo`.

This makes simple things wasy: `=>print($)`, `=>$1-$2`, `=>"${$}"` (OK, that one sucks a little).
It makes more complicated things possible: `=>42`, `listOfLists.map(=>$.map(=>$$.toString()).join("")).join("")`.

We can then still consider allowing *trailing function literals*: When a function invocation's last argument is a function literal, `foo(a1, a2, (...) { ... })`, you can write it as `foo(a1, a2) { ... }` instead. 
Maybe even allow you to drop the trailing semicolon if the expression has type `void` (or require it to have type void, since it looks like a loop constructor which doesn't have a value - yet).

Then you can do:
```dart
  list.forEach() {
    print($1);
  }
```

**Comment by eernstg** [comment]
I like the `$1 .. $k` and `$named` idea! It allows us to rely on a larger set of context types, and it reads quite reasonably.

For the trailing function literals, we still have the issue that `foo(a1, a2) { ... }` is a local function declaration, and we would presumably have to parse some tokens following `}` in order to see whether it is actually a local function or it is an expression, e.g., `foo(a1, a2) {...};` or `foo(a1, a2) {...}  - 1`.

**Comment by ykmnkmi** [comment]
@eernstg what about `.1 .. .2` and `.named`? Of course, this may conflict with enums in the future.

**Comment by eernstg** [comment]
Right, and it may also conflict with another kind of concise function literal: `myList.forEach(.toString())` meaning `myList.forEach((x) => x.toString())`

**Comment by lrhn** [comment]
I don't think I'll ever accept a syntax for functions which does not have a delimiter. The `myList.forEach(.toString())` is too ambiguous. Take `myListOfLists.forEach(.map(.toString()))`, how will that be parsed - and why?
So, not particularly worried about breaking that.

It's true that simply doing `foo(a, b) { body; }` is ambiguous grammar, and I definitely don't want a trailing semicolon to be the decider, that's just unreadable and easy to get wrong. We might need more punctuation. ("Punctuation, is? fun!")

As for `.1`, `.2` ... those are currently valid double literals, and even if not, the syntax makes me think of member access, and there is no object to access (except some implicit "parameter object").

**Comment by tatumizer** [comment]
Speaking of a  dollar idea: in string interpolation, `"$1"` is currently an error. The syntax can be extended to allow references to the $-parameters:
```dart
print("param=$1, param2=$2, sum=${$1+$2}"); 
```
Last time when the $-names were discussed (in the context of tuples), the argument was raised about the awkwardness of using these vars in string interpolation. (this wasn't the only argument though).

**Comment by Levi-Lesches** [comment]
> `=>=>$1+$$1`  

> `=>=>$foo+$$foo`  

> `=>print($)`

> `=>"${$}"`

> `listOfLists.map(=>$.map(=>$$.toString()).join("")).join("")`

Just to keep perspective here, this can get out of hand very quickly. Closures themselves are syntax sugar for regular functions, and now we're into sugar-for-sugar territory. I think as closures get more complicated (and nested), we should encourage devs to break them out into simple, separate functions. 

Here's an example where the readability is about the same, but closures make it very clear what's happening:
```dart
void printNested(List<List> list) {
  list.forEach(=>=>print($$));
  // vs
  list.forEach(=>$.forEach(print));
  // vs
  list.forEach( (List sublist) => sublist.forEach(print) );
}

void main() {
  List<List> list = [[1, 2, 3], [4, 5, 6]];
  printNested(list);  // each number on its own line
}
```

And here's an example of a separate function being more readable than all the alternatives:
```dart
String groupNested1(List<List> list) => list
  .map(=>$.map(=>$$.toString().join("")))
  .join(", ");

// vs

String groupNested2(List<List> list) => list.map(
  (List sublist) => sublist.map( 
    (value) => value.toString()
  ).join("")
).join(", ");

// vs
  
String groupNested3(List<List> list) => [
  for (final List sublist in list) [
    for (final int value in sublist)
      value.toString()
  ].join("")
].join(", ");

// vs

String toString(Object obj) => obj.toString();
String groupList(List list) => list.map(toString).join("");
String groupNested4(List<List> list) => list.map(groupList).join(", ");

// --------

void main() {
  var list = [[1, 2, 3], [4, 5, 6]];
  print(groupNested(list));  // 123, 456
}
```

**Comment by gosoccerboy5** [comment]
Is removing the calling parentheses when an abbreviated function literal being considered?
```dart
myList.where { return it.even; };
```

**Comment by lrhn** [comment]
@gosoccerboy5 It's a feature of other languages that we are aware of. Whether it's a good fit for Dart depends on whether we can fit it into the existing syntax. If we can, that would be great.

**Comment by eernstg** [comment]
This is about syntax like `myWhile (someCondition) {/*myLambda, playing the role of a while body*/}`, and all the other variants where the point is that we allow the last parameter to be passed outside the `(...)` (and we may allow the `()` to be emitted entirely if there is only that last parameter), when the last parameter is of the form `{...}`.

The problem with that syntax in Dart is that it does _not_ use a consistent syntactic pattern where declarations start with a reserved word. (For instance, Scala does that, so they use `var` and `val` to introduce mutable/immutable variables, and it's always easy for the parser to recognize a variable declaration from the very first token).

So if we see `myWhile(b) { doSomething(); }` then it could be an invocation of a function named `myWhile` with the arguments `b` and `{ doSomething(); }`, and it could also be a declaration of a local function (whose parameter doesn't have a declared type).

**Comment by gosoccerboy5** [comment]
> ....
> This means that it isn't something that just works.

Oh.

**Comment by Wdestroier** [comment]
Could \`expr\` (an expression inside backticks) be a syntax sugar for `(it) => expr`? Example:

```java
xs.forEach(`print(it)`);
```

**Comment by eernstg** [comment]
> Could \`expr\` [mean] `(it) => expr`?

That could probably work (backticks are currently not used at all in Dart, so they are very easy for a parser to recognize). They wouldn't support nesting, but abbreviated function literals inside abbreviated function literals aren't going to be very useful anyway.

Backticks could be somewhat more concise than braces, because they wouldn't "naturally" imply spaces. Also, the `<nonEmptyBlock>` that I proposed would contain a list of statements, not an expression, and that would allow us to avoid `return` in a number of situations. Finally, we should compare these forms with the one that uses a plain `=>`:

```dart
xs.forEach(`print(it)`);
xs.forEach({ print(it); });
xs.forEach(=> print(it));

xs.map(`2 * it`);
xs.map({ return 2 * it; });
xs.map(=> 2 * it);
```

**Comment by lrhn** [comment]
I think nested abbreviated function literals can be useful, like `listOfLists.forEach(=>it.forEach(=>print("[$it]")))`.

(I'm leaning more towards the parameters of the function being derived from the context type, types, positions and names, and treating the parameter list like an implicit record type, so `void foo(int Function(int x, {required bool increment}) act) ...` would allow `foo(=>$1 + (increment ? 1 : -1))`.
Then we can special case the one-positional-argument case and call it `it`. And positional record fields should start at `$1` for this.)

**Comment by tjarvstrand** [comment]
Arguably, this is a testament to how much Dart as a language has evolved and improved over the years (great job, btw!), but as it stands, this is my single biggest gripe with it right now. Would be great to see something done here!
